src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToAsyncMethods.cs

```diff
@@ -0,0 +1,244 @@
+                {
+                    if (methodBody.SemanticModel.GetDeclaredSymbol(methodBody.Syntax) is IMethodSymbol method)
+                    {
+                        declaration = method;
+                        break;
+                    }
+                }
+                else if (currentOperation.Kind == OperationKind.Block && currentOperation is IBlockOperation methodBaseBody)
+                {
+                    if (methodBaseBody.SemanticModel.GetDeclaredSymbol(methodBaseBody.Syntax) is IMethodSymbol method)
+                    {
+                        declaration = method;
+                        // There are many kinds of blocks, so only break if we found a method symbol for this block.
+                        // Otherwise, blocks inside anonymous or local functions would not be detected - those would be the parent of the current operation.
+                        break;
+                    }
+                }
+
+                currentOperation = currentOperation.Parent;
+            }
+
+            return declaration != null;
+        }
+
+        // Check if the method only takes one ct and is the last parameter in the method signature.
+        // We want to compare the current method signature to any others with the exact same arguments in the exact same order.
+        private static bool VerifyMethodOnlyHasOneCancellationTokenAsLastArgument(
+            INamedTypeSymbol cancellationTokenType,
+            IMethodSymbol methodDeclaration,
+            out string? cancellationTokenParameterName)
+        {
+            cancellationTokenParameterName = null;
+
+            if (methodDeclaration.Parameters.Count(x => x.Type.Equals(cancellationTokenType)) == 1 &&
+                methodDeclaration.Parameters.Last() is IParameterSymbol lastParameter &&
+                lastParameter.Type.Equals(cancellationTokenType))
+            {
+                cancellationTokenParameterName = lastParameter.Name;
+            }
+
+            return cancellationTokenParameterName != null;
+        }
+
+        // Check if the currently used overload is the one that takes the ct, but is utilizing the default value offered in the method signature.
+        // We want to offer a diagnostic for this case, so the user explicitly passes the ancestor's ct.
+        private static bool InvocationIgnoresOptionalCancellationToken(INamedTypeSymbol cancellationTokenType, IInvocationOperation invocation, IMethodSymbol method)
+        {
+            if (method.Parameters.Any() && method.Parameters.Last() is IParameterSymbol lastParameter && lastParameter.Type.Equals(cancellationTokenType))
+            {
+                // If the ct parameter has a default value, return true if a value is not being explicitly passed in the invocation
+                return lastParameter.IsOptional && // Has a default value
+                       invocation.Arguments.Last() is IArgumentOperation lastArgument &&
```

Comment:
Last argument does not necessarily match the last parameter in presence of named arguments. You should instead iterate through all arguments to find the `IArgumentOperation` whose `.Parameter` matches `lastParameter`

Reply:
Might also be good to define an extension method in `IOperationExtensions`:

```csharp
public static IArgumentOperation? GetArgumentForParameter(this IInvocationOperation invocation, IParameterSymbol parameter)
    => invocation.Arguments.FirstOrDefault(a => a.Parameter.Equals(parameter));
```

Reply:
> _Last argument does not necessarily match the last parameter in presence of named arguments._

Correct. It's one of the items I added to the "Potential future improvements" section in the main description. I've been thinking more about this and I think we really want named parameters to be handled on the first version I merge. I'll get this addressed.

Reply:
> _Might also be good to define an extension method in IOperationExtensions_

Good idea. I think StreamAsyncOverloads analyzer I merged recently could consume it as well.

Reply:
I ended up not needing to do this.